File: C/CLibrary/List.c
#include "List.h"

#include <stddef.h>

/**
 * Compares the one specified list with another for equality.
 *
 * @param ptr The one list to be compared for equality
 * @param sizePtr The list size function
 * @param getPtr The list get function
 * @param o The other list to be compared for equality
 * @param sizeO The other list size function
 * @param getO The other list get function
 * @param equals The list element equals function
 * @return true if the one specified list is equal to the other
 */
bool equalsList(void *ptr, int32_t (*sizePtr)(void *), void *(*getPtr)(void *, int32_t), 
        void *o, int32_t (*sizeO)(void *), void *(*getO)(void *, int32_t), 
        bool (*equals)(const void *, const void *)) {
    if (ptr == o) {
        return true;
    }
    
    if (o == NULL) {
        return false;
    }
    
    int length = sizePtr(ptr);
    
    if (sizeO(o) != length) {
        return false;
    }
    
    int i;
    for (i = 0; i < length; i++) {
        void *o1 = getPtr(ptr, i);
        void *o2 = getO(o, i);
        bool e = o1 == NULL ? o2 == NULL : equals(o1, o2);
        
        if (!e) {
            return false;
        }
    }
    
    return true;
}

File: CSharp/CsLibrary/List.cs
namespace JavaClasses
{

    /// <summary>
    /// An ordered collection (also known as a sequence).
    /// </summary>
    /// <typeparam name="E">the type of elements in this list</typeparam>
    /// <author>Petr Kozler (A13B0359P)</author>
    public abstract class List<E> : Object where E : Object
    {
        
        public abstract bool add(E e);
        
        public abstract void add(int index, E element);

        /// <summary>
        /// Compares the specified object with this list for equality.
        /// </summary>
        /// <param name="o">the object to be compared for equality with this list</param>
        /// <returns>true if the specified object is equal to this list</returns>
        public override bool equals(Object o)
        {
            // comparing references
            if (this == o)
            {
                return true;
            }

            // testing the references for a NULL value
            if (o == null)
            {
                return false;
            }

            List<E> other = o as List<E>;
            int length = size();

            // comparing the list sizes
            if (other.size() != length)
            {
                return false;
            }

            // comparing the corresponding elements of the lists
            for (int i = 0; i < length; i++)
            {
                E o1 = get(i);
                E o2 = other.get(i);
                bool equals = o1 == null ? o2 == null : o1.Equals(o2);

                if (!equals)
                {
                    return false;
                }
            }

            return true;
        }
        
        public abstract E get(int index);
        
        public abstract E set(int index, E element);
        
        public abstract E remove(int index);
        
        public abstract int size();

        /// <summary>
        /// Returns true if this list contains no elements.
        /// </summary>
        /// <returns>true if this list contains no elements
        /// </returns>
        public virtual bool isEmpty()
        {
            return size() == 0;
        }
        
        public abstract void clear();

    }

}
